File: 动态规划/cpp/superEggDrop.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>

using namespace std;

// 时间效率为O(K*N^2),空间效率为O(KN)
int superEggDrop(int k, int N)
{
    if(k == 1)
        return N;
    if(N == 0)
        return 0;
    int res = INT32_MAX;
    for(int i=1; i<N+1; i++)
    {
        int broken_egg = superEggDrop(k-1, i-1);
        int not_broken = superEggDrop(k, N-i);
        int tmp = max({broken_egg, not_broken}) + 1;
        if(res > tmp)
            res = tmp;
    }
    return res;
}

int superEggDropBinSearch(int k, int N)
{
    if(k == 1)
        return N;
    if(N == 0)
        return 0;
    int res = INT32_MAX;
    int left = 1;
    int right = N + 1;
    while(left < right)
    {
        int mid = left + (right - left) / 2;
        int broken_egg = superEggDropBinSearch(k-1, mid-1);
        int not_broken = superEggDropBinSearch(k, N-mid);

        if (broken_egg < not_broken)
        {
            left = mid + 1;
            res = min({res, not_broken + 1});
        }
        else if (broken_egg > not_broken)
        {
            right = mid;
            res = min({res, broken_egg + 1});
        }
        else if (broken_egg == not_broken)
        {
            return broken_egg + 1;
        }
        
    }
    return res;
}

int main()
{   
    int k = 2;
    cout << superEggDropBinSearch(k, 10) << endl;
    return 0;
}

File: DynamicProgramming/python/superEggDrop.py
def super_egg_drop(K, N):
    """
    :param K: 鸡蛋数
    :param N: 楼层数
    :return: 找到鸡蛋恰好摔不碎的那层楼, 在最坏情况, 所需要的最少尝试次数
    """
    def dp(K, N):
        if K == 1:
            return N
        if N == 0:
            return 0

        res = float('inf')
        for i in range(1, N+1):
            res = min(
                res,
                max(dp(K-1, i-1), dp(K, N-i-1)) + 1,
                      )
        return res
    return dp(K, N)


def super_egg_drop_table(K, N):
    """
    时间复杂度: O(K*N^2), 空间复杂度O(KN)
    """
    d = {}
    def dp(K, N):
        if K == 1:
            return N
        if N == 0:
            return 0
        if (K, N) in d:
            return d[(K, N)]
        res = float('inf')
        for i in range(1, N+1):
            res = min(
                res,
                max(dp(K-1, i-1), dp(K, N-i-1)) + 1,    # 最坏情况下, 所以取max
                      )
        d[(K, N)] = res
        return res
    return dp(K, N)


def super_egg_drop_table_bin_search(K, N):
    """
    使用二分查找来改进算法, 时间复杂度: O(K*N*logN), 空间复杂度O(KN)
    """
    d = {}
    def dp(K, N):
        if K == 1:
            return N
        if N == 0:
            return 0
        if (K, N) in d:
            return d[(K, N)]
        res = float('inf')
        lo = 1
        hi = N
        while lo <= hi:
            mid = (lo + hi) // 2
            broken = dp(K-1, mid-1)
            not_broken = dp(K, N-mid-1)
            if broken > not_broken:
                hi = mid - 1
                res = min(res, broken + 1)
            else:
                lo = mid + 1
                res = min(res, not_broken + 1)

        d[(K, N)] = res
        return res
    return dp(K, N)


# def super_egg_drop_itertion(K, N):
#     """
#     迭代版本, 时间复杂度: O(K*N), 空间复杂度O(KN)
#     """
#     import numpy as np
#     dp_table = np.zeros((K+1, N+1))
#     m = 0
#     while dp_table[K, m] < N:
#         m += 1
#         for k in range(1, K+1):
#             dp_table[k, m] = dp_table[k, m-1] + dp_table[k-1, m-1] + 1
#     return m


if __name__ == "__main__":
    print(super_egg_drop_table_bin_search(2, 10))
